core/src/main/java/org/apache/spark/shuffle/sort/ShuffleInMemorySorter.java

```diff
@@ -69,7 +72,11 @@ public int numRecords() {
     return pos;
   }
 
-  public void reset() {
+  public void freeMemoryAndResetPointerArray() {
+    if (consumer != null) {
+      consumer.freeArray(array);
+      this.array = consumer.allocateArray(initialSize);
```

Comment:
This should never fail because at this point, the task should not be holding any memory and we are just allocating the pointer array of initial size, right? 

Also, `allocateArray` will internally call `allocatePage` of the  `TaskMemoryManager` which grabs the lock anyway. So grabbing that lock here wont make sense.